csrc/fusion_profiler.h

```diff
@@ -0,0 +1,266 @@
+  double effective_bandwidth_gbs{0.0};
+  double percentage_peak_bandwidth{0.0};
+
+  std::array<int32_t, 3> grid{0, 0, 0};
+  std::array<int32_t, 3> block{0, 0, 0};
+  std::array<uint32_t, 3> cluster{0, 0, 0};
+
+  int32_t dynamic_shared_mem{0};
+  int32_t static_shared_mem{0};
+  uint32_t registers{0};
+
+  int64_t input_bytes{0};
+  int64_t output_bytes{0};
+
+  std::string device_name{};
+  double peak_bandwidth_gbs{0.0};
+};
+
+//! \struct FusionProfile
+//! \brief This struct captures the profiled information from Fusion that
+//! includes aggregated times from the kernels generated by segments
+//! encapsulated by a Fusion.
+struct FusionProfile {
+  //! A static array to capture header strings for tables that print
+  //! the profiled information
+  static std::array<const char*, 25> column_strs;
+
+  void reset();
+
+  bool verbose{isProfilerPrintingVerbose()};
+  int64_t fusion_id{-1};
+  int64_t segments{0};
+
+  double cuda_evt_time_ms{0.0};
+  double host_time_ms{0.0};
+  double compile_time_ms{0.0};
+  double kernel_time_ms{0.0};
+
+  int64_t input_bytes{0};
+  int64_t output_bytes{0};
+
+  double effective_bandwidth_gbs{0.0};
+  double percentage_peak_bandwidth{0.0};
+
+  //! Vector of of the KernelProfiles for each segment of a Fusion
+  std::vector<KernelProfile> kernel_profiles{};
+};
+
+std::ostream& operator<<(std::ostream&, const FusionProfile&);
+
+//! \struct SegmentProfiler
+//! \brief A class used to profile each segment of a Fusion
+class SegmentProfiler {
+ public:
+  SegmentProfiler(uint32_t id, bool cupti_disabled);
+
+  void startCompile(int device);
+  void stopCompile();
+
+  void startKernel(int device);
+  void stopKernel();
+
+  void inputBytesAccessed(int64_t bytes);
+  void outputBytesAccessed(int64_t bytes);
+
+  uint32_t segmentId() const;
+  int device() const {
+    return device_;
+  }
+
+  int64_t inputBytes() const {
+    return input_bytes_;
+  }
+  int64_t outputBytes() const {
+    return output_bytes_;
+  }
+  double compileTime() {
+    return compile_timer_.time();
+  }
+  ProfilerState state() const {
+    return kernel_profile_state_;
+  }
+
+ private:
+  bool cupti_disabled_;
+
+  int device_;
+  uint32_t segment_id_;
+
+  HostTimer compile_timer_;
+  int64_t input_bytes_;
+  int64_t output_bytes_;
+  ProfilerState kernel_profile_state_;
+};
+
+//! \struct FusionProfiler
+//! \brief A singleton class to profile Fusions that can include multiple
+//! segments.
+class FusionProfiler {
+  FusionProfiler();
+
+ public:
+  //! Method to access FusionProfiler singleton
+  static FusionProfiler* get();
+
+  void reset();
+  ProfilerState state() const;
+  SegmentProfiler& segment(size_t idx);
+
+  //! Profiling Methods
+  void start(bool cupti_disable = false);
+  void stop();
+  void createSegments(size_t num);
+  void startCompile();
+  void stopCompile();
+  void inputBytesAccessed(int64_t bytes);
+  void outputBytesAccessed(int64_t bytes);
+  const FusionProfile& profile() const;
```

Comment:
Could these all be made `static` and just call a private `get()` under the hood? Then we'll never be tempted to hold onto a `FusionProfiler*` since we won't be able to construct one.

Reply:
I can do that.  You are right there is no real reason to hold onto the `FusionProfiler*`.